### ISSUE TITLE
FormDialog, LuisDialog and multiple choice validation ignores first selection

### ISSUE THREAD

**Issue by itoledo** [Microsoft/BotBuilder#1900]
I have a very strange issue.
I have a LuisDialog with many LuisIntents that works fine. In one of my intents, I receive my entity and then call a FormDialog:

```
[LuisIntent("ubicacion_usuario")]
        public async Task ProcesarUbicacionUsuario(IDialogContext context, LuisResult result)
        {
            Debug.WriteLine("ProcesarUbicacionUsuario");
            EntityRecommendation lugar;
            var entities = new List<EntityRecommendation>(result.Entities);
            result.TryFindEntity("lugar", out lugar);
            if (lugar != null)
            {
                Debug.WriteLine($"detectado entity: {lugar.Entity}");
                entities.Add(new EntityRecommendation(type: "Lugar") { Entity = lugar.Entity });
            }
            var dlg = new FormDialog<LugarData>(new LugarData(), BuildSetLugarForm, FormOptions.PromptInStart, entities);
            context.Call(dlg, SetLugarComplete);
        }

        [Serializable]
        public class LugarData
        {
            [Describe("Tu ubicación")]
            [Prompt("Dónde estás ahora?")]
            [Template(TemplateUsage.NotUnderstood, "Al parecer \"{0}\" no es un lugar válido.")]
            public string Lugar;
        }

        private static IForm<LugarData> BuildSetLugarForm()
        {
            var builder = new FormBuilder<LugarData>();
            return builder
                .Field(nameof(LugarData.Lugar),
                    validate: ValidaLugar)
                .Build();
        }

```
My validation function calls a webservice to validate addresses, which calls a webservice. If it returns a single address, validation is OK. If it returns many, it asks the user for disambiguation.

```
public static async Task<ValidateResult> ValidaLugar(object state, object response)
        {
            Debug.WriteLine("ValidarLugar: " + response as string);
            ValidateResult valresult = new ValidateResult() { IsValid = true, Value = response };
            try
            {
                var str = response as string;
                if (str.StartsWith("json:"))
                {
                    var lugar = JsonConvert.DeserializeObject<Lugar>(str.Substring(5));
                    if (lugar != null)
                        return valresult;
                }
            }
            catch (Exception)
            {
                Debug.WriteLine($"excepcion al deserializar lugar: {response}");
            }

            var ls = await SearchService.SearchNominatimPlaceFromText(response as string);
            if (ls.Count == 0)
            {
                valresult.IsValid = false;
                valresult.Feedback = "No encontramos ese lugar";
            }
            else if (ls.Count > 1)
            {
                valresult.IsValid = false;
                valresult.Feedback = "Encontramos más de un lugar, por favor elige:";
                var cs = new List<Choice>();
                foreach (var l in ls)
                {
                    Debug.WriteLine($"alternativa: {l.nombre}");
                    cs.Add(new Choice()
                    {
                        Description =
                            new DescribeAttribute(l.nombre + ", " + l.comuna, ResponseBuilderService.BuildURLForPlace(l)),
                        Terms = new TermsAttribute(l.nombre + ", " + l.comuna),
                        Value = "json:" + JsonConvert.SerializeObject(l)
                    });
                }
                valresult.Choices = cs;
            }
            else
                valresult.Value = "json:" + JsonConvert.SerializeObject(ls[0]);

            return valresult;
        }

```
Problem is, when there are choices, it displays them fine. But the first time the user picks an option, it seems that the answer is being ignored, and the bot just returns the FormDialog prompt ("Dónde estás ahora?"). The second time it works fine.

BTW I know about the new Bing Location form dialog but it does not work very well here in my country.

Any hints will be very much appreciated!

Thanks,
Ivan

**Comment by chrimc62** [comment]
I have tests on this very issue and it works in them.  Is it possible to share out your repository so I can take a look?

**Comment by itoledo** [comment]
@chrimc62 here it is... I guess the problem is in either Controller/MessagesController.cs or Dialogs/StritLuisDialog.cs.
Thanks!

[StritBot-20161228.zip](https://github.com/Microsoft/BotBuilder/files/675006/StritBot-20161228.zip)

**Comment by chrimc62** [comment]
Can you give me a script that causes the behavior you see?

**Comment by itoledo** [comment]
Try with:

- "estoy en torre entel" (this means "I'm at torre entel", which is a landmark place in Santiago de Chile)
- "voy a plaza Italia" (this means "I'm going to Plaza Italia", another landmark place)

the second line will trigger a multiple choice option. Selecting it the first time will give you the prompt. The second time it will work.

Thanks again!

**Comment by chrimc62** [comment]
This was indeed a bug which I have fixed and will be in the next release--thanks for reporting it!  (Most likely next week.)  The issue was that when processing passed in entities they are treated as "input" to each field so that you get clarification, etc.   If during that processing I get simple feedback like "Your address must start with a number" I do not immediately ask for that field--it is just asked in the normal flow.  In your case, you are returning buttons to help fill in the field, so I have changed the code that it will ask for that field to be filled in right then.
